aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu

```diff
@@ -153,6 +153,79 @@ __global__ void compute_grad_weight(
   grad_weight_per_segment[id * stride + startFeature] = weight;
 }
 
+// Fused kernel that combines compute_grad_weight and sum_and_scatter using atomic adds.
+// This eliminates the serialization bottleneck when num_of_segments is small.
+// Each partial segment atomically adds its contribution directly to grad_weight.
+template <typename scalar_t, typename index_t>
+__global__ void compute_grad_weight_fused(
+    const index_t *orig_indices,
+    const scalar_t *gradOutput,
+    const index_t *count,
+    ptrdiff_t numel,
+    int64_t stride,
+    const index_t *partial_segment_offsets,
+    const int64_t *num_of_partial_segments_ptr,
+    const index_t *sorted_indices,
+    const index_t *partial_to_segment_idx,
+    const index_t *segment_offsets,
+    scalar_t *grad_weight,
+    const int64_t padding_idx,
+    const int64_t stride_warped) {
+
+  int64_t num_of_partial_segments = *num_of_partial_segments_ptr;
+  using accscalar_t = acc_type<scalar_t, true>;
+  const int64_t gid = static_cast<int64_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int64_t partial_id = gid / stride_warped;
+  const int64_t startFeature = gid % stride_warped;
+
+  if (startFeature >= stride) {
+    return;
+  }
+  if (partial_id >= num_of_partial_segments) {
+    return;
+  }
+
+  const int idx_begin = partial_segment_offsets[partial_id];
+  const int idx_end = (partial_id == num_of_partial_segments - 1)
+      ? numel
+      : partial_segment_offsets[partial_id + 1];
+
+  accscalar_t weight = 0;
+  for (int idx = idx_begin; idx < idx_end; ++idx) {
+    const index_t target_row = orig_indices[idx];
+    const accscalar_t scale = count ? (accscalar_t)1.0 / count[idx] : 1.0;
+    weight += gradOutput[target_row * stride + startFeature] * scale;
+  }
+
+  // Get the target row for this partial segment from the segment info
+  const index_t segment_id = partial_to_segment_idx[partial_id];
+  const int64_t target_row = sorted_indices[segment_offsets[segment_id]];
+
+  if (target_row != padding_idx) {
+    gpuAtomicAddNoReturn(
```

Comment:
Are we now using atomics in a path that didn't use them, do we need to check if we are doing things in deterministic mode?